benchmarks/benchmark.cc

```diff
@@ -70,47 +71,90 @@ static void BM_ArenaInitialBlockOneAlloc(benchmark::State& state) {
-    bytes_per_iter = _upb_DefPool_BytesLoaded(symtab.ptr());
+enum LoadDescriptorMode {
+  NoLayout,
+  WithLayout,
+};
+
+// This function is mostly copied from upb/def.c, but it is modified to avoid
+// passing in the pre-generated mini-tables, in order to force upb to compute
+// them dynamically.  Generally you would never want to do this, but we want to
+// simulate the cost we would pay if we were loading these types purely from
+// descriptors, with no mini-tales available.
+bool LoadDefInit_NoLayout(upb_DefPool *s, const _upb_DefPool_Init *init,
+                          size_t *bytes) {
+  _upb_DefPool_Init** deps = init->deps;
+  google_protobuf_FileDescriptorProto* file;
+  upb_Arena* arena;
+  upb_Status status;
+
+  upb_Status_Clear(&status);
+
+  if (upb_DefPool_FindFileByName(s, init->filename)) {
+    return true;
   }
-  state.SetBytesProcessed(state.iterations() * bytes_per_iter);
+
+  arena = upb_Arena_New();
+
+  for (; *deps; deps++) {
+    if (!LoadDefInit_NoLayout(s, *deps, bytes))
+      goto err;
+  }
+
+  file = google_protobuf_FileDescriptorProto_parse_ex(
+      init->descriptor.data, init->descriptor.size, NULL,
+      kUpb_DecodeOption_AliasString, arena);
+  *bytes += init->descriptor.size;
+
+  if (!file) {
+    upb_Status_SetErrorFormat(
+        &status,
+        "Failed to parse compiled-in descriptor for file '%s'. This should "
+        "never happen.",
+        init->filename);
+    goto err;
+  }
+
+  // KEY DIFFERENCE: Here we pass in only the descriptor, and not the
+  // pre-generated minitables.
+  if (!upb_DefPool_AddFile(s, file, &status)) {
+    goto err;
+  }
+
+  upb_Arena_Free(arena);
+  return true;
+
+err:
+  fprintf(stderr,
+          "Error loading compiled-in descriptor for file '%s' (this should "
+          "never happen): %s\n",
+          init->filename, upb_Status_ErrorMessage(&status));
+  exit(1);
 }
-BENCHMARK(BM_LoadDescriptor_Upb);
 
+template <LoadDescriptorMode Mode>
 static void BM_LoadAdsDescriptor_Upb(benchmark::State& state) {
   size_t bytes_per_iter = 0;
   for (auto _ : state) {
     upb::SymbolTable symtab;
-    google_ads_googleads_v7_services_SearchGoogleAdsRequest_getmsgdef(
-        symtab.ptr());
-    bytes_per_iter = _upb_DefPool_BytesLoaded(symtab.ptr());
-  }
-  state.SetBytesProcessed(state.iterations() * bytes_per_iter);
-}
-BENCHMARK(BM_LoadAdsDescriptor_Upb);
-
-static void BM_LoadDescriptor_Proto2(benchmark::State& state) {
-  for (auto _ : state) {
-    protobuf::Arena arena;
-    protobuf::StringPiece input(descriptor.data, descriptor.size);
-    auto proto =
-        protobuf::Arena::CreateMessage<protobuf::FileDescriptorProto>(&arena);
-    protobuf::DescriptorPool pool;
-    bool ok = proto->ParseFrom<protobuf::MessageLite::kMergePartial>(input) &&
-              pool.BuildFile(*proto) != nullptr;
-    if (!ok) {
-      printf("Failed to add file.\n");
-      exit(1);
+    if (Mode == NoLayout) {
```

Comment:
is this boolean backwards?  It feels like you should call NoLayout when NoLayout is true

Reply:
I can see what that is confusing. I've renamed the function to `LoadDefInit_BuildLayout()` to make this more clear. The function forces a layout to be computed by not passing in the compiled-in layout. It was called `NoLayout` because it does not pass a layout in (forcing one to be computed), whereas the `NoLayout` benchmarks are benchmarking the case that a layout is not computed.